File: array/0034-firstlast.cpp
/*
 * leetcode 34
 * find first and last position of element in sorted array
 *
 * sorted -> binary search
 *
 */

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        int i = 0, j = n-1;
        int m = 0;
        vector<int> res(2, -1);
        bool found = false;

        // binary search
        while (i <= j) {
            m = i + (j-i)/2;
            if (nums[m] == target) {
                found = true;
                break;
            } else if (nums[m] > target) {
                j = m-1;
            } else {
                i = m+1;
            }
        }
        if (!found) {
            return res;
        }
        i = m;
        while((i-1) >= 0 && (nums[m]==nums[i-1])) {
            i--;
        }
        res[0] = i;
        j = m;
        while((j+1) < n && (nums[m]==nums[j+1])) {
            j++;
        }
        res[1] = j;
        return res;    
    }
};

//
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int i = 0, j = nums.size()-1; // [i, j]
        while (i <= j) {
            int m = i + (j-i)/2;
            if (nums[m] < target) {
                i = m+1; // [m+1, j]
            } else {
                j = m-1; // [i, m-1]
            }
        }
        return i;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        int start = search(nums, target);
        if (start == nums.size() || nums[start] != target)
            return {-1, -1};
        int end = search(nums, target+1) -1;
        return {start, end};
    }
};

File: array/0034-firstlast.py
"""
leetcode 34
binary search

"""

def lower_bound(nums: List[int], target: int) -> int:
    # [left, right]
    left, right = 0, len(nums) - 1
    while left <= right:
        # loop invariant
        # nums[left-1] < target
        # nums[right+1] >= target
        mid = (left + right) // 2
        if nums[mid] < target:
            left = mid + 1  # --> [mid+1, right]
        else:
            right = mid - 1  # --> [left, mid-1]
    return left

def lower_bound2(nums: List[int], target: int) -> int:
    left = 0
    right = len(nums)  # range: [left, right)
    while left < right:  # 
        # loop invariants:
        # nums[left-1] < target
        # nums[right] >= target
        mid = (left + right) // 2
        if nums[mid] < target:
            left = mid + 1  # range --> [mid+1, right)
        else:
            right = mid  # range --> [left, mid)
    return left  # or right

#
# (left, right)
#
def lower_bound3(nums: List[int], target: int) -> int:
    left, right = -1, len(nums)  # range: (left, right)
    while left + 1 < right:  # 
        mid = (left + right) // 2
        # loop invariants:
        # nums[left] < target
        # nums[right] >= target
        if nums[mid] < target:
            left = mid  # range --> (mid, right)
        else:
            right = mid  # range --> (left, mid)
    return right  # or left+1

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        start = lower_bound(nums, target)  # either one above
        if start == len(nums) or nums[start] != target:
            return [-1, -1]
        # if start exists, then end exists too.
        end = lower_bound(nums, target + 1) - 1
        return [start, end]

